File: Sharebook/Server/src/Sharebook.Streams/TrackedStream.cs
namespace WorldFeed.Streams
{
    using System;
    using System.Collections.Generic;
    using System.Linq;
    using System.Text;
    using System.Threading.Tasks;

    using WorldFeed.Common.Events;
    using WorldFeed.Common.Helpers;
    using WorldFeed.Common.Public;
    using WorldFeed.Common.Public.Client.Tools;
    using WorldFeed.Common.Public.Events;
    using WorldFeed.Common.Public.Models.Enums;
    using WorldFeed.Common.Public.Models.Interfaces;
    using WorldFeed.Common.Public.Parameters.StreamsClient;
    using WorldFeed.Common.Public.Streaming.Enums;
    using WorldFeed.Common.Streaming;
    using WorldFeed.Common.Wrappers;
    using WorldFeed.Streams.Properties;

    public class TrackedStream : TwitterStream, ITrackedStream
    {
        public event EventHandler<MatchedTweetReceivedEventArgs> TweetReceived;
        public event EventHandler<MatchedTweetReceivedEventArgs> MatchingTweetReceived;
        public event EventHandler<TweetEventArgs> NonMatchingTweetReceived;

        private readonly ITwitterClient client;
        private readonly IStreamTrackManager<ITweet> streamTrackManager;
        private readonly ITwitterClientFactories factories;

        public override event EventHandler<StreamEventReceivedArgs> EventReceived;

        public TrackedStream(
            ITwitterClient client,
            IStreamTrackManager<ITweet> streamTrackManager,
            IJsonObjectConverter jsonObjectConverter,
            IJObjectStaticWrapper jObjectStaticWrapper,
            IStreamResultGenerator streamResultGenerator,
            ITwitterClientFactories factories,
            ICreateTrackedTweetStreamParameters createTrackedTweetStreamParameters)

            : base(streamResultGenerator, jsonObjectConverter, jObjectStaticWrapper, createTrackedTweetStreamParameters)
        {
            this.client = client;
            this.streamTrackManager = streamTrackManager;
            this.factories = factories;
        }

        public async Task StartAsync(string url)
        {
            ITwitterRequest createTwitterRequest()
            {
                var queryBuilder = new StringBuilder(url);
                AddBaseParametersToQuery(queryBuilder);

                var request = this.client.CreateRequest();
                request.Query.Url = queryBuilder.ToString();
                request.Query.HttpMethod = HttpMethod.GET;
                return request;
            }

            void onJsonReceived(string json)
            {
                RaiseJsonObjectReceived(json);

                if (IsEvent(json))
                {
                    TryInvokeGlobalStreamMessages(json);
                    return;
                }

                var tweet = this.factories.CreateTweet(json);

                var detectedTracksAndActions = this.streamTrackManager.GetMatchingTracksAndActions(tweet.FullText);
                var detectedTracks = detectedTracksAndActions.Select(x => x.Item1);

                var eventArgs = new MatchedTweetReceivedEventArgs(tweet, json) { MatchingTracks = detectedTracks.ToArray(), };

                if (detectedTracksAndActions.Any())
                {
                    eventArgs.MatchOn = MatchOn.TweetText;

                    RaiseTweetReceived(eventArgs);
                    RaiseMatchingTweetReceived(eventArgs);
                }
                else
                {
                    RaiseTweetReceived(eventArgs);
                    RaiseNonMatchingTweetReceived(new TweetEventArgs(tweet, json));
                }
            }

            await this.streamResultGenerator.StartAsync(onJsonReceived, createTwitterRequest).ConfigureAwait(false);
        }

        protected void RaiseJsonObjectReceived(string json)
        {
            this.Raise(EventReceived, new StreamEventReceivedArgs(json));
        }

        public int TracksCount => this.streamTrackManager.TracksCount;
        public int MaxTracks => this.streamTrackManager.MaxTracks;
        public Dictionary<string, Action<ITweet>> Tracks => this.streamTrackManager.Tracks;

        public void AddTrack(string track, Action<ITweet> trackReceived = null)
        {
            if (this.streamResultGenerator.StreamState != StreamState.Stop)
            {
                throw new InvalidOperationException(Resources.TrackedStream_ModifyTracks_NotStoppedException_Description);
            }

            this.streamTrackManager.AddTrack(track, trackReceived);
        }

        public void RemoveTrack(string track)
        {
            if (this.streamResultGenerator.StreamState != StreamState.Stop)
            {
                throw new InvalidOperationException(Resources.TrackedStream_ModifyTracks_NotStoppedException_Description);
            }

            this.streamTrackManager.RemoveTrack(track);
        }

        public bool ContainsTrack(string track)
        {
            return this.streamTrackManager.ContainsTrack(track);
        }

        public void ClearTracks()
        {
            if (this.streamResultGenerator.StreamState != StreamState.Stop)
            {
                throw new InvalidOperationException(Resources.TrackedStream_ModifyTracks_NotStoppedException_Description);
            }

            this.streamTrackManager.ClearTracks();
        }

        protected void RaiseTweetReceived(MatchedTweetReceivedEventArgs eventArgs)
        {
            this.Raise(TweetReceived, eventArgs);
        }

        protected void RaiseMatchingTweetReceived(MatchedTweetReceivedEventArgs eventArgs)
        {
            this.Raise(MatchingTweetReceived, eventArgs);
        }

        protected void RaiseNonMatchingTweetReceived(TweetEventArgs eventArgs)
        {
            this.Raise(NonMatchingTweetReceived, eventArgs);
        }
    }
}

File: Sharebook/Client/src/app/streams/TrackedStream.ts
import {ITrackedStream} from "../core/Core/Streaming/ITrackedStream";
import {ITweet} from "../core/Public/Models/Interfaces/ITweet";
import {ITwitterClient} from "../core/Public/ITwitterClient";
import {ITwitterClientFactories} from "../core/Public/Client/Tools/ITwitterClientFactories";
import {ICreateTrackedTweetStreamParameters} from "../core/Public/Parameters/StreamsClient/CreateTrackedStreamParameters";
import {IStreamResultGenerator} from "../core/Core/Streaming/IStreamResultGenerator";
import {IStreamTrackManager} from "../core/Core/Streaming/IStreamTrackManager";
import {ITwitterRequest} from "../core/Public/Models/Interfaces/ITwitterRequest";
import StringBuilder from "../c#-objects/TypeScript.NET-Core/packages/Core/source/Text/StringBuilder";
import {HttpMethod} from "../core/Public/Models/Enum/HttpMethod";
import Dictionary from "../c#-objects/TypeScript.NET-Core/packages/Core/source/Collections/Dictionaries/Dictionary";
import {StreamResources} from "./Properties/stream-resources";
import InvalidOperationException from "../c#-objects/TypeScript.NET-Core/packages/Core/source/Exceptions/InvalidOperationException";
import {StreamState} from "../core/Public/Streaming/StreamState";
import {Action} from "../c#-objects/TypeScript.NET-Core/packages/Core/source/FunctionTypes";

export class TrackedStream extends TwitterStream implements ITrackedStream {
  public event EventHandler<MatchedTweetReceivedEventArgs> TweetReceived;
  public event EventHandler<MatchedTweetReceivedEventArgs> MatchingTweetReceived;
  public event EventHandler<TweetEventArgs> NonMatchingTweetReceived;

  private readonly _client: ITwitterClient;
  private readonly _streamTrackManager: IStreamTrackManager<ITweet>;
  private readonly _factories: ITwitterClientFactories;

  public override event EventHandler<StreamEventReceivedArgs> EventReceived;

  constructor(
    client: ITwitterClient,
    streamTrackManager: IStreamTrackManager<ITweet>,
    jsonObjectConverter: IJsonObjectConverter,
    jObjectStaticWrapper: IJObjectStaticWrapper,
    streamResultGenerator: IStreamResultGenerator,
    factories: ITwitterClientFactories,
    createTrackedTweetStreamParameters: ICreateTrackedTweetStreamParameters) {
    super(streamResultGenerator, jsonObjectConverter, jObjectStaticWrapper, createTrackedTweetStreamParameters);
    this._client = client;
    this._streamTrackManager = streamTrackManager;
    this._factories = factories;
  }

        public async  StartAsync(url: string): Promise<void>
        {
            ITwitterRequest createTwitterRequest()
            {
              let queryBuilder = new StringBuilder(url);
                AddBaseParametersToQuery(queryBuilder);

                let request = this._client.createRequest();
                request.query.url = queryBuilder.toString();
                request.query.httpMethod = HttpMethod.GET;
                return request;
            }

             onJsonReceived(json: string): void
            {
                RaiseJsonObjectReceived(json);

                if (IsEvent(json))
                {
                    TryInvokeGlobalStreamMessages(json);
                    return;
                }

                var tweet = _factories.CreateTweet(json);

                var detectedTracksAndActions = _streamTrackManager.GetMatchingTracksAndActions(tweet.FullText);
                var detectedTracks = detectedTracksAndActions.Select(x => x.Item1);

                var eventArgs = new MatchedTweetReceivedEventArgs(tweet, json) { MatchingTracks = detectedTracks.ToArray(), };

                if (detectedTracksAndActions.Any())
                {
                    eventArgs.MatchOn = MatchOn.TweetText;

                    RaiseTweetReceived(eventArgs);
                    RaiseMatchingTweetReceived(eventArgs);
                }
                else
                {
                    RaiseTweetReceived(eventArgs);
                    RaiseNonMatchingTweetReceived(new TweetEventArgs(tweet, json));
                }
            }

            await _streamResultGenerator.StartAsync(onJsonReceived, createTwitterRequest).ConfigureAwait(false);
        }

        protected  RaiseJsonObjectReceived(json: string): void
        {
            this.Raise(EventReceived, new StreamEventReceivedArgs(json));
        }


  get TracksCount(): number {
    return this._streamTrackManager.TracksCount;
  }

  get MaxTracks(): number {
    return this._streamTrackManager.MaxTracks;
  }

  get Tracks(): Dictionary<string, Action<ITweet>> {
    return this._streamTrackManager.Tracks;
  }

        public  AddTrack(track: string, trackReceived: Action<ITweet> = null): void
        {
            if (_streamResultGenerator.StreamState != StreamState.Stop)
            {
                throw new InvalidOperationException(Resources.TrackedStream_ModifyTracks_NotStoppedException_Description);
            }

            _streamTrackManager.AddTrack(track, trackReceived);
        }

        public  RemoveTrack(track: string): void
        {
            if (_streamResultGenerator.StreamState != StreamState.Stop)
            {
                throw new InvalidOperationException(StreamResources.TrackedStream_ModifyTracks_NotStoppedException_Description);
            }

            _streamTrackManager.RemoveTrack(track);
        }

        public  ContainsTrack(track: string): boolean
        {
            return _streamTrackManager.ContainsTrack(track);
        }

        public  ClearTracks(): void
        {
            if (_streamResultGenerator.StreamState != StreamState.Stop)
            {
                throw new InvalidOperationException(StreamResources.TrackedStream_ModifyTracks_NotStoppedException_Description);
            }

            _streamTrackManager.ClearTracks();
        }

        protected  RaiseTweetReceived(eventArgs: MatchedTweetReceivedEventArgs): void
        {
            this.Raise(TweetReceived, eventArgs);
        }

        protected  RaiseMatchingTweetReceived(eventArgs: MatchedTweetReceivedEventArgs): void
        {
            this.Raise(MatchingTweetReceived, eventArgs);
        }

        protected  RaiseNonMatchingTweetReceived(eventArgs: TweetEventArgs): void
        {
            this.Raise(NonMatchingTweetReceived, eventArgs);
        }
    }
}
